File: lib/game/regions/region_data.dart
import 'package:flutter/material.dart';

/// Which silhouette set the background painter should draw for a region.
enum Landmark {
  mumbaiSkyline,
  goaBeach,
  keralaBackwaters,
  bengaluruTech,
  chennaiTemple,
  hyderabadCharminar,
  jaipurFort,
  delhiGate,
  varanasiGhats,
  northeastHills,
  kashmirValley,
  ladakhPeaks,
}

class RegionData {
  final String id;
  final String name;
  final String tagline;
  final Landmark landmark;

  /// Sky gradient, top to bottom.
  final Color skyTop;
  final Color skyBottom;

  /// Silhouette / midground colors, back to front (parallax layers).
  final List<Color> layerColors;

  final Color groundColor;
  final Color accentColor;

  /// Flavor names used for obstacle labels / collectible names — keeps
  /// every region mechanically identical but thematically distinct.
  final List<String> obstacleFlavors;
  final String collectibleName;
  final String powerUpVehicle;
  final String chaseFlavor;

  /// Cumulative distance (meters) at which this region begins.
  final double unlockDistance;

  const RegionData({
    required this.id,
    required this.name,
    required this.tagline,
    required this.landmark,
    required this.skyTop,
    required this.skyBottom,
    required this.layerColors,
    required this.groundColor,
    required this.accentColor,
    required this.obstacleFlavors,
    required this.collectibleName,
    required this.powerUpVehicle,
    required this.chaseFlavor,
    required this.unlockDistance,
  });
}

/// The journey: west coast -> south -> deccan -> north -> Himalaya finale.
class Regions {
  static const List<RegionData> all = [
    RegionData(
      id: 'mumbai',
      name: 'Mumbai',
      tagline: 'Catch the local before it leaves without you',
      landmark: Landmark.mumbaiSkyline,
      skyTop: Color(0xFF3B5670),
      skyBottom: Color(0xFFE8A97D),
      layerColors: [Color(0xFF24384A), Color(0xFF35506A), Color(0xFF4A6B88)],
      groundColor: Color(0xFF2E2E33),
      accentColor: Color(0xFFFFC857),
      obstacleFlavors: ['Local train barrier', 'Auto-rickshaw', 'Vada pav cart', 'Monsoon puddle'],
      collectibleName: 'Vada Pav',
      powerUpVehicle: 'Auto-rickshaw Dash',
      chaseFlavor: 'the last local pulling out of the platform',
      unlockDistance: 0,
    ),
    RegionData(
      id: 'goa',
      name: 'Goa',
      tagline: 'Sun, sand and a scooter with no brakes',
      landmark: Landmark.goaBeach,
      skyTop: Color(0xFF4FA6D8),
      skyBottom: Color(0xFFFFE29A),
      layerColors: [Color(0xFF1E6E8C), Color(0xFF2E93AE), Color(0xFFF2C879)],
      groundColor: Color(0xFFE8C48A),
      accentColor: Color(0xFFFF6F59),
      obstacleFlavors: ['Beach shack', 'Sun lounger', 'Parked scooter', 'Coconut stand'],
      collectibleName: 'Cashew Feni Shell',
      powerUpVehicle: 'Scooter Sprint',
      chaseFlavor: 'the incoming high tide',
      unlockDistance: 900,
    ),
    RegionData(
      id: 'kerala',
      name: 'Kerala',
      tagline: 'Backwaters, houseboats and a thousand palms',
      landmark: Landmark.keralaBackwaters,
      skyTop: Color(0xFF5FA88C),
      skyBottom: Color(0xFFDCEFC7),
      layerColors: [Color(0xFF1F5C4A), Color(0xFF2E7D5F), Color(0xFF4FA37B)],
      groundColor: Color(0xFF3B6B4F),
      accentColor: Color(0xFFE8B93E),
      obstacleFlavors: ['Houseboat mast', 'Coconut tree fall', 'Fishing net', 'Canal gate'],
      collectibleName: 'Banana Leaf Coin',
      powerUpVehicle: 'Shikara Glide',
      chaseFlavor: 'the rising backwater current',
      unlockDistance: 1800,
    ),
    RegionData(
      id: 'bengaluru',
      name: 'Bengaluru',
      tagline: 'Namma metro, endless traffic, and deadlines',
      landmark: Landmark.bengaluruTech,
      skyTop: Color(0xFF6B7A99),
      skyBottom: Color(0xFFC9D3E0),
      layerColors: [Color(0xFF2B3350), Color(0xFF3D4A72), Color(0xFF576E9E)],
      groundColor: Color(0xFF33363D),
      accentColor: Color(0xFF3ED6B5),
      obstacleFlavors: ['Traffic barricade', 'Food delivery bike', 'Glass tower scaffold', 'Metro pillar'],
      collectibleName: 'Filter Coffee Token',
      powerUpVehicle: 'Metro Dash',
      chaseFlavor: 'a gridlock wave swallowing the road behind you',
      unlockDistance: 2700,
    ),
    RegionData(
      id: 'chennai',
      name: 'Chennai',
      tagline: 'Marina winds and temple gopurams',
      landmark: Landmark.chennaiTemple,
      skyTop: Color(0xFF3E7CA6),
      skyBottom: Color(0xFFF7DCA0),
      layerColors: [Color(0xFF224C63), Color(0xFF2F6B87), Color(0xFF6FA8C4)],
      groundColor: Color(0xFFD8C79A),
      accentColor: Color(0xFFE0533D),
      obstacleFlavors: ['Temple cart', 'Beach kite string', 'Bus stop queue', 'Fishing boat'],
      collectibleName: 'Filter Kaapi Tumbler',
      powerUpVehicle: 'Cycle-rickshaw Rush',
      chaseFlavor: 'a cyclone gust rolling off the Marina',
      unlockDistance: 3600,
    ),
    RegionData(
      id: 'hyderabad',
      name: 'Hyderabad',
      tagline: 'Bazaars, biryani steam and Irani chai',
      landmark: Landmark.hyderabadCharminar,
      skyTop: Color(0xFFB0568F),
      skyBottom: Color(0xFFF3B562),
      layerColors: [Color(0xFF5B2A4C), Color(0xFF7C3B63), Color(0xFFAE5D82)],
      groundColor: Color(0xFF4A3A2E),
      accentColor: Color(0xFFF6C453),
      obstacleFlavors: ['Bazaar stall', 'Biryani handi cart', 'Pearl shop crate', 'Metro pillar'],
      collectibleName: 'Osmania Biscuit',
      powerUpVehicle: 'Metro Dash',
      chaseFlavor: 'the bazaar crowd closing in behind you',
      unlockDistance: 4500,
    ),
    RegionData(
      id: 'jaipur',
      name: 'Jaipur',
      tagline: 'Forts, camels and a rising dust storm',
      landmark: Landmark.jaipurFort,
      skyTop: Color(0xFFD98A4A),
      skyBottom: Color(0xFFF6D9A0),
      layerColors: [Color(0xFF7A4A2D), Color(0xFFA5643A), Color(0xFFD08B4F)],
      groundColor: Color(0xFFC98F4E),
      accentColor: Color(0xFFE8546B),
      obstacleFlavors: ['Camel cart', 'Market stall', 'Fort gate barrier', 'Puppet stand'],
      collectibleName: 'Lac Bangle',
      powerUpVehicle: 'Camel Dash',
      chaseFlavor: 'a wall of desert dust (the loo)',
      unlockDistance: 5400,
    ),
    RegionData(
      id: 'delhi',
      name: 'Delhi',
      tagline: 'Metro lines, India Gate and Chandni Chowk lanes',
      landmark: Landmark.delhiGate,
      skyTop: Color(0xFF7A8AA6),
      skyBottom: Color(0xFFE7D9C4),
      layerColors: [Color(0xFF3A4257), Color(0xFF52607F), Color(0xFF8593AC)],
      groundColor: Color(0xFF474747),
      accentColor: Color(0xFFDA6A2E),
      obstacleFlavors: ['Metro barrier', 'Street food cart', 'Cycle rickshaw', 'Wedding baraat'],
      collectibleName: 'Parantha Token',
      powerUpVehicle: 'Metro Dash',
      chaseFlavor: 'rush-hour traffic surging behind you',
      unlockDistance: 6300,
    ),
    RegionData(
      id: 'varanasi',
      name: 'Varanasi',
      tagline: 'Ghats, narrow lanes and the evening aarti',
      landmark: Landmark.varanasiGhats,
      skyTop: Color(0xFFB6784F),
      skyBottom: Color(0xFFF4D9A6),
      layerColors: [Color(0xFF5C3A2E), Color(0xFF7C4F3B), Color(0xFFA9744F)],
      groundColor: Color(0xFF8C7259),
      accentColor: Color(0xFFFF9E3D),
      obstacleFlavors: ['Ghat steps', 'Boat mooring rope', 'Diya seller', 'Narrow lane cow'],
      collectibleName: 'Brass Diya',
      powerUpVehicle: 'Boat Glide',
      chaseFlavor: 'the Ganga aarti flame wave',
      unlockDistance: 7200,
    ),
    RegionData(
      id: 'northeast',
      name: 'Northeast India',
      tagline: 'Living root bridges, tea gardens and misty hills',
      landmark: Landmark.northeastHills,
      skyTop: Color(0xFF4E7A6B),
      skyBottom: Color(0xFFBFDCC8),
      layerColors: [Color(0xFF264439), Color(0xFF356554), Color(0xFF4E8B72)],
      groundColor: Color(0xFF3D5C46),
      accentColor: Color(0xFF9FD8C0),
      obstacleFlavors: ['Root bridge gap', 'Tea garden basket', 'Bamboo scaffold', 'River crossing'],
      collectibleName: 'Tea Leaf Basket',
      powerUpVehicle: 'Cable Crossing',
      chaseFlavor: 'monsoon mist rolling down the valley',
      unlockDistance: 8100,
    ),
    RegionData(
      id: 'kashmir',
      name: 'Kashmir',
      tagline: 'Snow, mountains and a shikara on the lake',
      landmark: Landmark.kashmirValley,
      skyTop: Color(0xFF8FB6D9),
      skyBottom: Color(0xFFEAF3FA),
      layerColors: [Color(0xFF4B6D8C), Color(0xFF6E93B4), Color(0xFFAFCBE0)],
      groundColor: Color(0xFFEFF4F8),
      accentColor: Color(0xFFE0553F),
      obstacleFlavors: ['Snow drift', 'Houseboat plank', 'Shikara oar', 'Chinar branch'],
      collectibleName: 'Kahwa Cup',
      powerUpVehicle: 'Shikara Glide',
      chaseFlavor: 'an avalanche rumbling down the slope',
      unlockDistance: 9000,
    ),
    RegionData(
      id: 'ladakh',
      name: 'Ladakh',
      tagline: 'The roof of the journey — high passes and prayer flags',
      landmark: Landmark.ladakhPeaks,
      skyTop: Color(0xFF1E3A5F),
      skyBottom: Color(0xFF8FB8D6),
      layerColors: [Color(0xFF2C3E52), Color(0xFF3F5871), Color(0xFF6C8AA3)],
      groundColor: Color(0xFFB79E7E),
      accentColor: Color(0xFFE84C3D),
      obstacleFlavors: ['Prayer flag pole', 'Mountain boulder', 'Frozen stream', 'Monastery step'],
      collectibleName: 'Prayer Flag',
      powerUpVehicle: 'Yak Dash',
      chaseFlavor: 'a mountain blizzard closing the pass',
      unlockDistance: 9900,
    ),
  ];

  static RegionData forDistance(double meters) {
    RegionData current = all.first;
    for (final r in all) {
      if (meters >= r.unlockDistance) current = r;
    }
    return current;
  }

  static int indexForDistance(double meters) {
    int idx = 0;
    for (int i = 0; i < all.length; i++) {
      if (meters >= all[i].unlockDistance) idx = i;
    }
    return idx;
  }
}

File: unity/Assets/Scripts/Data/RegionData.cs
using System.Collections.Generic;
using UnityEngine;

namespace Gati.Data
{
    /// <summary>
    /// One stop on the journey across India. Plain data — no MonoBehaviour —
    /// so it can be used from gameplay code, terrain generation and UI alike.
    /// </summary>
    public class RegionData
    {
        public string Id;
        public string DisplayName;
        public string Tagline;
        public PropKind Landmark;

        public Color SkyTop;
        public Color SkyBottom;
        public Color GroundColor;
        public Color AccentColor;
        public bool SnowGround;

        public string[] ObstacleFlavors;
        public string CollectibleName;
        public string PowerUpVehicle;
        public string ChaseFlavor;

        /// Cumulative distance (meters) at which this region begins.
        public float UnlockDistance;
    }

    /// <summary>
    /// The journey, west coast -> south -> deccan -> north -> Himalaya finale.
    /// Mirrors lib/game/regions/region_data.dart from the original Flutter
    /// prototype so the two versions stay conceptually in sync.
    /// </summary>
    public static class Regions
    {
        public static readonly List<RegionData> All = new List<RegionData>
        {
            new RegionData
            {
                Id = "mumbai", DisplayName = "Mumbai",
                Tagline = "Catch the local before it leaves without you",
                Landmark = PropKind.MumbaiSkyline,
                SkyTop = new Color32(0x3B, 0x56, 0x70, 255), SkyBottom = new Color32(0xE8, 0xA9, 0x7D, 255),
                GroundColor = new Color32(0x2E, 0x2E, 0x33, 255), AccentColor = new Color32(0xFF, 0xC8, 0x57, 255),
                ObstacleFlavors = new[] { "Local train barrier", "Auto-rickshaw", "Vada pav cart", "Monsoon puddle" },
                CollectibleName = "Vada Pav", PowerUpVehicle = "Auto-rickshaw Dash",
                ChaseFlavor = "the last local pulling out of the platform", UnlockDistance = 0,
            },
            new RegionData
            {
                Id = "goa", DisplayName = "Goa",
                Tagline = "Sun, sand and a scooter with no brakes",
                Landmark = PropKind.GoaPalms,
                SkyTop = new Color32(0x4F, 0xA6, 0xD8, 255), SkyBottom = new Color32(0xFF, 0xE2, 0x9A, 255),
                GroundColor = new Color32(0xE8, 0xC4, 0x8A, 255), AccentColor = new Color32(0xFF, 0x6F, 0x59, 255),
                ObstacleFlavors = new[] { "Beach shack", "Sun lounger", "Parked scooter", "Coconut stand" },
                CollectibleName = "Cashew Feni Shell", PowerUpVehicle = "Scooter Sprint",
                ChaseFlavor = "the incoming high tide", UnlockDistance = 900,
            },
            new RegionData
            {
                Id = "kerala", DisplayName = "Kerala",
                Tagline = "Backwaters, houseboats and a thousand palms",
                Landmark = PropKind.KeralaBackwaters,
                SkyTop = new Color32(0x5F, 0xA8, 0x8C, 255), SkyBottom = new Color32(0xDC, 0xEF, 0xC7, 255),
                GroundColor = new Color32(0x3B, 0x6B, 0x4F, 255), AccentColor = new Color32(0xE8, 0xB9, 0x3E, 255),
                ObstacleFlavors = new[] { "Houseboat mast", "Coconut tree fall", "Fishing net", "Canal gate" },
                CollectibleName = "Banana Leaf Coin", PowerUpVehicle = "Shikara Glide",
                ChaseFlavor = "the rising backwater current", UnlockDistance = 1800,
            },
            new RegionData
            {
                Id = "bengaluru", DisplayName = "Bengaluru",
                Tagline = "Namma metro, endless traffic, and deadlines",
                Landmark = PropKind.BengaluruTech,
                SkyTop = new Color32(0x6B, 0x7A, 0x99, 255), SkyBottom = new Color32(0xC9, 0xD3, 0xE0, 255),
                GroundColor = new Color32(0x33, 0x36, 0x3D, 255), AccentColor = new Color32(0x3E, 0xD6, 0xB5, 255),
                ObstacleFlavors = new[] { "Traffic barricade", "Food delivery bike", "Glass tower scaffold", "Metro pillar" },
                CollectibleName = "Filter Coffee Token", PowerUpVehicle = "Metro Dash",
                ChaseFlavor = "a gridlock wave swallowing the road behind you", UnlockDistance = 2700,
            },
            new RegionData
            {
                Id = "chennai", DisplayName = "Chennai",
                Tagline = "Marina winds and temple gopurams",
                Landmark = PropKind.ChennaiTemple,
                SkyTop = new Color32(0x3E, 0x7C, 0xA6, 255), SkyBottom = new Color32(0xF7, 0xDC, 0xA0, 255),
                GroundColor = new Color32(0xD8, 0xC7, 0x9A, 255), AccentColor = new Color32(0xE0, 0x53, 0x3D, 255),
                ObstacleFlavors = new[] { "Temple cart", "Beach kite string", "Bus stop queue", "Fishing boat" },
                CollectibleName = "Filter Kaapi Tumbler", PowerUpVehicle = "Cycle-rickshaw Rush",
                ChaseFlavor = "a cyclone gust rolling off the Marina", UnlockDistance = 3600,
            },
            new RegionData
            {
                Id = "hyderabad", DisplayName = "Hyderabad",
                Tagline = "Bazaars, biryani steam and Irani chai",
                Landmark = PropKind.HyderabadCharminar,
                SkyTop = new Color32(0xB0, 0x56, 0x8F, 255), SkyBottom = new Color32(0xF3, 0xB5, 0x62, 255),
                GroundColor = new Color32(0x4A, 0x3A, 0x2E, 255), AccentColor = new Color32(0xF6, 0xC4, 0x53, 255),
                ObstacleFlavors = new[] { "Bazaar stall", "Biryani handi cart", "Pearl shop crate", "Metro pillar" },
                CollectibleName = "Osmania Biscuit", PowerUpVehicle = "Metro Dash",
                ChaseFlavor = "the bazaar crowd closing in behind you", UnlockDistance = 4500,
            },
            new RegionData
            {
                Id = "jaipur", DisplayName = "Jaipur",
                Tagline = "Forts, camels and a rising dust storm",
                Landmark = PropKind.JaipurFort,
                SkyTop = new Color32(0xD9, 0x8A, 0x4A, 255), SkyBottom = new Color32(0xF6, 0xD9, 0xA0, 255),
                GroundColor = new Color32(0xC9, 0x8F, 0x4E, 255), AccentColor = new Color32(0xE8, 0x54, 0x6B, 255),
                ObstacleFlavors = new[] { "Camel cart", "Market stall", "Fort gate barrier", "Puppet stand" },
                CollectibleName = "Lac Bangle", PowerUpVehicle = "Camel Dash",
                ChaseFlavor = "a wall of desert dust (the loo)", UnlockDistance = 5400,
            },
            new RegionData
            {
                Id = "delhi", DisplayName = "Delhi",
                Tagline = "Metro lines, India Gate and Chandni Chowk lanes",
                Landmark = PropKind.DelhiGate,
                SkyTop = new Color32(0x7A, 0x8A, 0xA6, 255), SkyBottom = new Color32(0xE7, 0xD9, 0xC4, 255),
                GroundColor = new Color32(0x47, 0x47, 0x47, 255), AccentColor = new Color32(0xDA, 0x6A, 0x2E, 255),
                ObstacleFlavors = new[] { "Metro barrier", "Street food cart", "Cycle rickshaw", "Wedding baraat" },
                CollectibleName = "Parantha Token", PowerUpVehicle = "Metro Dash",
                ChaseFlavor = "rush-hour traffic surging behind you", UnlockDistance = 6300,
            },
            new RegionData
            {
                Id = "varanasi", DisplayName = "Varanasi",
                Tagline = "Ghats, narrow lanes and the evening aarti",
                Landmark = PropKind.VaranasiGhats,
                SkyTop = new Color32(0xB6, 0x78, 0x4F, 255), SkyBottom = new Color32(0xF4, 0xD9, 0xA6, 255),
                GroundColor = new Color32(0x8C, 0x72, 0x59, 255), AccentColor = new Color32(0xFF, 0x9E, 0x3D, 255),
                ObstacleFlavors = new[] { "Ghat steps", "Boat mooring rope", "Diya seller", "Narrow lane cow" },
                CollectibleName = "Brass Diya", PowerUpVehicle = "Boat Glide",
                ChaseFlavor = "the Ganga aarti flame wave", UnlockDistance = 7200,
            },
            new RegionData
            {
                Id = "northeast", DisplayName = "Northeast India",
                Tagline = "Living root bridges, tea gardens and misty hills",
                Landmark = PropKind.NortheastHills,
                SkyTop = new Color32(0x4E, 0x7A, 0x6B, 255), SkyBottom = new Color32(0xBF, 0xDC, 0xC8, 255),
                GroundColor = new Color32(0x3D, 0x5C, 0x46, 255), AccentColor = new Color32(0x9F, 0xD8, 0xC0, 255),
                ObstacleFlavors = new[] { "Root bridge gap", "Tea garden basket", "Bamboo scaffold", "River crossing" },
                CollectibleName = "Tea Leaf Basket", PowerUpVehicle = "Cable Crossing",
                ChaseFlavor = "monsoon mist rolling down the valley", UnlockDistance = 8100,
            },
            new RegionData
            {
                Id = "kashmir", DisplayName = "Kashmir",
                Tagline = "Snow, mountains and a shikara on the lake",
                Landmark = PropKind.KashmirValley,
                SkyTop = new Color32(0x8F, 0xB6, 0xD9, 255), SkyBottom = new Color32(0xEA, 0xF3, 0xFA, 255),
                GroundColor = new Color32(0xEF, 0xF4, 0xF8, 255), AccentColor = new Color32(0xE0, 0x55, 0x3F, 255),
                SnowGround = true,
                ObstacleFlavors = new[] { "Snow drift", "Houseboat plank", "Shikara oar", "Chinar branch" },
                CollectibleName = "Kahwa Cup", PowerUpVehicle = "Shikara Glide",
                ChaseFlavor = "an avalanche rumbling down the slope", UnlockDistance = 9000,
            },
            new RegionData
            {
                Id = "ladakh", DisplayName = "Ladakh",
                Tagline = "The roof of the journey — high passes and prayer flags",
                Landmark = PropKind.LadakhPeaks,
                SkyTop = new Color32(0x1E, 0x3A, 0x5F, 255), SkyBottom = new Color32(0x8F, 0xB8, 0xD6, 255),
                GroundColor = new Color32(0xB7, 0x9E, 0x7E, 255), AccentColor = new Color32(0xE8, 0x4C, 0x3D, 255),
                SnowGround = true,
                ObstacleFlavors = new[] { "Prayer flag pole", "Mountain boulder", "Frozen stream", "Monastery step" },
                CollectibleName = "Prayer Flag", PowerUpVehicle = "Yak Dash",
                ChaseFlavor = "a mountain blizzard closing the pass", UnlockDistance = 9900,
            },
        };

        public static RegionData ForDistance(float meters)
        {
            RegionData current = All[0];
            foreach (var r in All)
            {
                if (meters >= r.UnlockDistance) current = r;
            }
            return current;
        }
    }
}
